### PR TITLE
Tickets/dm 43722

### PR DESCRIPTION
**Pull request by rcboufleur**
#### Overview
This pull request implements a structured framework for processing and transforming data from the Engineering and Facilities Database (EFD). It enables data retrieval, transformation, schema generation, and integration within the LSST ecosystem.

Main components:
- Configurable data transformation of EFD topics
- Automated schema generation for database tables
- Task scheduling and retry management
- Statistical summarization of time-series data
- YAML-based configuration files per instrument
- Customized Alembic migration versioning for schema updates

This implementation supports data processing for instruments such as LATISS, LSSTComCam, and LSSTComCamSim.

##### Configuration Framework
- Pydantic-based models (`Field`, `Topic`, `Column`, `ConfigModel`) for YAML validation
- Validation for unpivoted tables
- Instrument-specific configurations (e.g., `config_latiss.yaml`)

##### Data Transformation Pipeline
- Processes exposures and visits using the `Transform` class
- Queries EFD data via `InfluxDbDao`
- Dynamically processes columns based on configuration
- Stores transformed data in PostgreSQL via `ExposureEfdDao/VisitEfdDao`
- Supports structured (pivoted) and key-value (unpivoted) data formats

##### Summary Statistics
- Computes metrics such as mean, standard deviation, and RMS from time-series data
- Supports time-based filtering (e.g., most recent value in the last minute)
- Integrates with transformation pipelines

##### Schema Generation & Alembic Migrations
- `generate_schema_from_config.py` creates database schemas from configuration files
- Supports structured tables (`exposure_efd`, `visit1_efd`) and key-value tables (`exposure_efd_unpivoted`, `visit1_efd_unpivoted`)
- Includes `transformed_efd_scheduler` table for task tracking
- Implements customized Alembic migration versioning:
  - Saves and restores schema snapshots during migrations
  - Tracks configuration changes across database versions

##### Task Management
- `QueueManager` handles task creation, retries, and status tracking
- Supports execution via Kubernetes Jobs (fixed intervals) and CronJobs (scheduled runs)
- Implements automatic retries for failed tasks with defined limits

#### Code Structure
- `config_model.py` – Configuration validation models
- `summary.py` – Statistical operations on time-series data
- `transform.py` – Core transformation logic
- `transform_efd.py` – CLI entry point and workflow orchestration
- `generate_schema_from_config.py` – Schema generation
- `dao/*.py` – Database access layer (PostgreSQL/InfluxDB)
- `queue_manager.py` – Task queue management

#### Validation & Error Handling
- **Configuration validation:** Ensures valid tables for unpivoted data and correct data types in summaries
- **Task retry logic:** Automatic retries with detailed error logging
- **Data integrity checks:** Time range validation for exposures/visits and database constraint enforcement
- **Schema migration validation:** Ensures Alembic migration consistency with stored snapshots

#### Testing
- **Unit tests:** Validation of configuration models, statistical calculations, and transformation edge cases (incomplete)

### CODE REVIEW

File: .vscode/settings.json
Comment (JeremyMcCormick) [inline .vscode/settings.json:1]: Can you remove this?

File: python/lsst/consdb/transformed_efd/dao/base.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+import numpy
+import pandas
+from sqlalchemy import Engine, MetaData, Table, create_engine
+from sqlalchemy import exc as sa_exc
+from sqlalchemy import func
+from sqlalchemy.dialects import postgresql, sqlite
+from sqlalchemy.pool import NullPool
+
+
+class DBBase:
```
Comment (JeremyMcCormick) [inline python/lsst/consdb/transformed_efd/dao/base.py:15]: Are there any other common utilities that could be used for something like this in ConsDB? (generic DAO implementation)

File: .vscode/settings.json
Comment (ktlim) [reply .vscode/settings.json:1]: Should `.vscode/` be added to `.gitignore` as well?

File: Dockerfile.efdtransform [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# Define the LSST version
+ARG OBS_LSST_VERSION=w_2024_47
+FROM lsstsqre/centos:7-stack-lsst_distrib-${OBS_LSST_VERSION}
```
Comment (ktlim) [inline Dockerfile.efdtransform:3]: Should change now to `ghcr.io/lsst/scipipe:al9-${OBS_LSST_VERSION}`, with the version at `w_2025_11` or so.

File: Dockerfile.efdtransform [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    pip install \
+    kafkit==0.2.1 \
+    lsst_efd_client==0.12.0
+
+# Copy the Python module
+COPY --chown=lsst:lsst python /opt/lsst/software/stack/python
+
+# Create and populate the data directory
+RUN mkdir -p /opt/lsst/software/stack/data
+COPY --chown=lsst:lsst tmp/efd_transform/*.db /opt/lsst/software/stack/data/
```
Comment (ktlim) [inline Dockerfile.efdtransform:19]: I'm not seeing where this `tmp` directory comes from. Should this be removed for production?

File: Dockerfile.efdtransform [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+# Environment variables
+ENV CONFIG_FILE="/opt/lsst/software/stack/python/lsst/consdb/efd_transform/config_LATISS.yml" \
```
Comment (ktlim) [inline Dockerfile.efdtransform:22]: This and `INSTRUMENT` will have to be overridden for LSSTCam deployments, I presume.

I think this is easier to maintain as a sequence of ENV lines rather than using backslashes.

File: Dockerfile.efdtransform [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    INSTRUMENT="LATISS" \
+    BUTLER_REPO="s3://rubin-summit-users/butler.yaml" \
+    S3_ENDPOINT_URL="https://s3dfrgw.slac.stanford.edu/" \
+    LSST_RESOURCES_S3_PROFILE_embargo="https://sdfembs3.sdf.slac.stanford.edu" \
+    PGUSER="rubin" \
+    CONSDB_URL="sqlite:////opt/lsst/software/stack/data/test.db" \
+    TIMEDELTA="5" \
+    LOG_FILE="/opt/lsst/software/stack/data/transform.log" \
```
Comment (ktlim) [inline Dockerfile.efdtransform:30]: Writing to stdout or stderr is more Kubernetes-friendly, and it doesn't risk (as much) filling the disk and crashing the job.  But if jobs deal with short time periods, this is not so much of an issue.

File: pyproject.toml [hunk 6]
```diff
@@ -44,4 +44,4 @@ asyncio_mode = "auto"
 test = ["pytest"]
 dev = [
     "documenteer[guide] < 2",
-]
+]
```
Comment (ktlim) [inline pyproject.toml:47]: I think the original with the proper line termination is better here.

File: python/lsst/consdb/transformed_efd/config_model.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
+"""Defines the configuration models used in the application."""
```
Comment (ktlim) [inline python/lsst/consdb/transformed_efd/config_model.py:1]: Please include the standard headers: https://developer.lsst.io/stack/license-and-copyright.html

File: python/lsst/consdb/transformed_efd/__init__.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+"""Provides a structured framework for processing and transforming data from the (EFD).
```
Comment (ktlim) [inline python/lsst/consdb/transformed_efd/__init__.py:1]: The documentation here does not exactly conform to https://developer.lsst.io/python/numpydoc.html though it is close.  A later ticket can be used to clean this up.

**Review by ktlim — COMMENTED** [review]
Still working through this, but posting some initial thoughts.

File: .vscode/settings.json
Comment (rcboufleur) [reply .vscode/settings.json:1]: Sure, I'll fix that. It should actually already be in `.gitignore`.

File: Dockerfile.efdtransform (hunk 3 above)
Comment (rcboufleur) [reply Dockerfile.efdtransform:19]: Yes, it does. It was the database used before we had postgres available.

File: Dockerfile.efdtransform (hunk 4 above)
Comment (rcboufleur) [reply Dockerfile.efdtransform:22]: Yes, all env variables can be overridden by Kubernetes manifests or direct function call.
I will change the backslashes by ENV lines.

File: Dockerfile.efdtransform (hunk 5 above)
Comment (rcboufleur) [reply Dockerfile.efdtransform:30]: It currently writes logs to both stdout/stderr and a log file. The log file can be removed or made optional.

File: alembic/transformed_efd_latiss/env.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+def save_config_snapshot(revision):
+    """Save config_latiss.yaml snapshot for a migration."""
+    snapshot_path = os.path.join(SNAPSHOT_DIR, f"{revision}_config_latiss.yaml")
+    shutil.copy(CONFIG_LATISS_PATH, snapshot_path)
+    logger.info(f"Saved config snapshot: {snapshot_path}")
+
+
+# Function to restore config snapshot and create a backup
+def restore_config_snapshot(revision):
+    """Restore config_latiss.yaml from a migration snapshot and create a backup."""
```
[the remark this answers is not in the document]
Comment (rcboufleur) [reply alembic/transformed_efd_latiss/env.py:72]: Yes, I have to check why I am having different flake8 and ruff rulesets on GitHub workflows.

File: python/lsst/consdb/transformed_efd/config_model.py (hunk 7 above)
Comment (rcboufleur) [reply python/lsst/consdb/transformed_efd/config_model.py:1]: Acknowledged, I'll implement that.

File: python/lsst/consdb/transformed_efd/dao/base.py (hunk 1 above)
Comment (rcboufleur) [reply python/lsst/consdb/transformed_efd/dao/base.py:15]: I am not sure.

File: python/lsst/consdb/transformed_efd/dao/base.py [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+        -------
+            The database engine.
+
+        """
+        if self.engine is None:
+            self.engine = create_engine(self.db_uri, poolclass=NullPool)
+
+        return self.engine
+
+    def get_con(self):
```
[the remark this answers is not in the document]
Comment (rcboufleur) [reply python/lsst/consdb/transformed_efd/dao/base.py:72]: Sure. It could make its reusability more efficient. I'll check that out.

**Comment by JeremyMcCormick** [comment]
Should the large config files in this PR go in another repo?

There is an existing one called last-dm/consdb_config which is maybe for this kind of use but it is currently empty.

**Comment by JeremyMcCormick** [comment]
@rcboufleur 

> +103,143

I would like to bring up a general issue on this pull request, which is that there is an extremely large amount of config being added (over 100k lines) to the repo that seems as if it could be unmanageable going forward.

I would like to understand what you are doing here in the implementation so that possible alternatives could be explored. It does not seem to me that having this amount of configuration is desirable or perhaps even necessary.

**Comment by rcboufleur** [comment]
@JeremyMcCormick 

The configuration files have become quite extensive but this is a direct result of the flexibility needed to customize inputs without modifying the code. They enable tailored functions, parametrizations, topics, and fields for each transformed quantity. With some instruments containing over 400 columns, a single configuration file can easily exceed 15,000 lines. These files also double as the foundation for generating the database schemas. In essence, their purpose is to minimize code adjustments when adding new quantities or removing existing ones. That said, ideas on how to optimize this are welcome.

**Comment by JeremyMcCormick** [comment]
> @JeremyMcCormick
> 
> The configuration files have become quite extensive but this is a direct result of the flexibility needed to customize inputs without modifying the code. They enable tailored functions, parametrizations, topics, and fields for each transformed quantity. With some instruments containing over 400 columns, a single configuration file can easily exceed 15,000 lines. These files also double as the foundation for generating the database schemas. In essence, their purpose is to minimize code adjustments when adding new quantities or removing existing ones. That said, ideas on how to optimize this are welcome.

Thanks for the explanation. 

A couple points of discussion:

1. Perhaps these files should be managed in a separate repository and checked out as a git submodule.
2. Would it make sense to have them stored in a document database or similar system? Having something like 100k lines of config on the filesystem seems to me pretty unwieldy, and managing this "by hand" going forward could be difficult.
3. Could we have a group design session where we discuss optimizations? Perhaps for instance some of these similarly named, numbered columns could be specified with wildcards.

**Comment by bbrondel** [comment]
I think a better strategy would be to (1) have this configuration generated on-the-fly, or at least have code that generates the configuration and reproduces the existing configuration files exactly, and then (2) add additional human-edited configuration files as needed, which would be manageable in size and which would override the automatically generated configuration.

**Comment by JeremyMcCormick** [comment]
> I think a better strategy would be to (1) have this configuration generated on-the-fly, or at least have code that generates the configuration and reproduces the existing configuration files exactly, and then (2) add additional human-edited configuration files as needed, which would be manageable in size and which would override the automatically generated configuration.

Something like this sounds more reasonable to me than storing such a giant (and basically unmanageable) amount of config in the git repo. It isn't clear to me how this config is generated though in terms of what has been written by hand and what has been generated by scripts/code.

**Comment by JeremyMcCormick** [comment]
For files like [python/lsst/consdb/transformed_efd/config/config_latiss.yaml](https://github.com/lsst-dm/consdb/pull/72/files#diff-882495961ac349293ec422d9f882a940df7d82bc7137fb4abddf002b4a9ade28), I would suggest to drop the `config_` prepend as that is already implicit in the file's path.

**Comment by JeremyMcCormick** [comment]
For columns in the Felis YAML files, the description field should be filled and possibly also `ivoa:ucd` so that a generated documentation site is more user-friendly. You may want to include options for these in the `config` file format describing the column transformations.

**Comment by JeremyMcCormick** [comment]
You do not need `nullable: True` in the Felis YAML files, as this is the default in the Pydantic model. I would suggest to only include this field if it is `nullable: False`.

**Comment by JeremyMcCormick** [comment]
You can rebase this PR to just a few or even a single commit. (Up to you but the granular log of development changes and fixes does not seem that useful.)

File: Dockerfile.efdtransform (hunk 5 above)
Comment (JeremyMcCormick) [reply Dockerfile.efdtransform]: > The log file can be removed or made optional.

I would remove this entirely and just access the logs through `kubectl`.

File: python/lsst/consdb/transformed_efd/config_model.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    """
+
+    name: str
+    fields: List[Field]
+
+
+class Column(BaseModel):
+    """Represents a column with a name and optional tables.
+
+    Attributes
```
Comment (JeremyMcCormick) [inline python/lsst/consdb/transformed_efd/config_model.py:66]: Documentation for most of these attributes is missing.

**Review by ktlim — APPROVED** [review]
